src/System.Linq.Expressions/src/System/Linq/Expressions/LambdaExpression.cs

```diff
@@ -80,7 +69,30 @@ ReadOnlyCollection<ParameterExpression> parameters
         /// Gets the value that indicates if the lambda expression will be compiled with
         /// tail call optimization. 
         /// </summary>
-        public bool TailCall => _tailCall;
+        public bool TailCall => TailCallCore;
+
+        protected virtual bool TailCallCore => false;
+
+        [ExcludeFromCodeCoverage] // Unreachable
+        internal virtual ReadOnlyCollection<ParameterExpression> GetOrMakeParameters()
+        {
+            throw ContractUtils.Unreachable;
+        }
+
+        [ExcludeFromCodeCoverage] // Unreachable
+        public virtual ParameterExpression GetParameter(int index)
```

Comment:
Note that `IParameterProvider` is internal, but I have to mark these members as `public`. They can be omitted from the API contract though. Alternatively, we can make `IParameterProvider` public (just like `IArgumentProvider`). That may not be a bad option, because it enables users overriding `VisitLambda<T>` to avoid accessing `Parameters` and causing the read-only collection allocation. If we go down this path (after API review), we should also implement the interface on `CatchBlock` and `BlockExpression`. We can then likely also use this interface in places where we track scoping in visitors.